backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/CachedLibraries.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package org.jetbrains.kotlin.backend.konan
+
+import org.jetbrains.kotlin.konan.file.File
+import org.jetbrains.kotlin.konan.library.KonanLibrary
+import org.jetbrains.kotlin.konan.target.CompilerOutputKind
+import org.jetbrains.kotlin.konan.target.KonanTarget
+import org.jetbrains.kotlin.library.uniqueName
+import org.jetbrains.kotlin.utils.addToStdlib.firstNotNullResult
+
+internal class CachedLibraries(
+        private val target: KonanTarget,
+        allLibraries: List<KonanLibrary>,
+        explicitCaches: Map<KonanLibrary, String>,
+        implicitCacheDirectories: List<String>
+) {
+
+    class Cache(val kind: Kind, val path: String) {
+        enum class Kind { DYNAMIC, STATIC }
+    }
+
+    private val allCaches: Map<KonanLibrary, Cache> = allLibraries.mapNotNull { library ->
+        val explicitPath = explicitCaches[library]
+
+        val cache = if (explicitPath != null) {
+            val kind = when {
+                explicitPath.endsWith(target.family.dynamicSuffix) -> Cache.Kind.DYNAMIC
+                explicitPath.endsWith(target.family.staticSuffix) -> Cache.Kind.STATIC
+                else -> error("unexpected cache: $explicitPath")
+            }
+            Cache(kind, explicitPath)
+        } else {
+            implicitCacheDirectories.firstNotNullResult { path ->
+                val dir = File(path).takeIf { it.isDirectory } ?: return@firstNotNullResult null
+                val baseName = "${library.uniqueName}-cache"
+                val dynamicFile = dir.child(getArtifactName(baseName, CompilerOutputKind.DYNAMIC_CACHE))
+                val staticFile = dir.child(getArtifactName(baseName, CompilerOutputKind.STATIC_CACHE))
+
+                if (dynamicFile.exists) {
+                    Cache(Cache.Kind.DYNAMIC, dynamicFile.absolutePath)
+                } else if (staticFile.exists) {
+                    Cache(Cache.Kind.STATIC, staticFile.absolutePath)
+                } else {
+                    null
+                }
+            }
+        }
+
+        cache?.let { library to it }
+    }.toMap()
+
+    private fun getArtifactName(baseName: String, kind: CompilerOutputKind) =
+            "${kind.prefix(target)}$baseName${kind.suffix(target)}"
+
+    fun isLibraryCached(library: KonanLibrary): Boolean =
+            getLibraryCache(library) != null
+
+    fun getLibraryCache(library: KonanLibrary): Cache? =
+            allCaches[library]
+
+    val hasStaticCaches = allCaches.values.any {
```

Comment:
This value can be set earlier when you iterating over `allLibraries`. On the other hand, this will make code less clean.

Reply:
> This value can be set earlier when you iterating over allLibraries.

What for?

> On the other hand, this will make code less clean.

Exactly.